File: src/main/java/club/mindtech/mindbot/MindBot.java
package club.mindtech.mindbot;

import club.mindtech.mindbot.commands.Commands;
import club.mindtech.mindbot.events.AnnotatedEventListener;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.JDABuilder;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.hooks.AnnotatedEventManager;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.security.auth.login.LoginException;

public class MindBot {
    public static final Logger LOGGER = LoggerFactory.getLogger(MindBot.class);
    private static JDA API;

    public static void main(String[] args) {
        if (args.length == 0) {
            LOGGER.error("No token passed. Please pass token as an argument");
            System.exit(1);
        }
        final String token = args[0];
        createJDA(token);
        registerCommands();
    }

    private static void createJDA(String token) {
        try {
            API = JDABuilder
                .createDefault(token)
                .setEventManager(new AnnotatedEventManager())
                .addEventListeners(new AnnotatedEventListener())
                .build()
                .awaitReady();
        } catch (LoginException e) {
            LOGGER.error("Invalid Token. Please check your token");
            System.exit(1);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        LOGGER.info("JDA created");
    }

    public static JDA getAPI() {
        return API;
    }

    private static void registerCommands() {
        getAPI().getGuilds().forEach(MindBot::registerGuildCommands);
    }

    private static void registerGuildCommands(Guild guild) {
        guild.updateCommands()
            .addCommands(Commands.getSlashCommandData())
            .queue();
    }

    public static MongoDatabase getDatabase() {
        return database;
    }
}

File: src/main/java/club/mindtech/mindbot/MindBot.kt
package club.mindtech.mindbot

import club.mindtech.mindbot.commands.getSlashCommandData
import club.mindtech.mindbot.database.initDatabase
import club.mindtech.mindbot.events.InteractionListener
import club.mindtech.mindbot.util.env
import com.mongodb.client.MongoDatabase
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.JDABuilder
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.hooks.AnnotatedEventManager
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import kotlin.system.exitProcess

val log: Logger = LoggerFactory.getLogger(MindBot::class.java)

fun main() {
    MindBot(env("DISCORD_TOKEN"))
}

class MindBot(token: String) {
    companion object {
        lateinit var jda: JDA
        lateinit var db: MongoDatabase
    }

    init {
        jda = createJDA(token)
        registerCommands()
        db = initDatabase(env("DB_URL"), env("DB_NAME"))
    }

    private fun createJDA(token: String): JDA {
        try {
            return JDABuilder
                .createDefault(token)
                .setEventManager(AnnotatedEventManager())
                .addEventListeners(InteractionListener())
                .build()
                .awaitReady()
        } catch (e: Exception) {
            log.error("Failed to create JDA: {}", e.message)
            exitProcess(1)
        }
    }

    private fun registerCommands() {
        jda.guilds.forEach {
            registerGuildCommands(it)
        }
    }

    private fun registerGuildCommands(guild: Guild) {
        guild.updateCommands()
            .addCommands(*getSlashCommandData())
            .queue()
    }
}
